Add Extensions to Gloo Validation: Rate Limit Configs

jackstine opened the pull request:
# Description

Add Extensions to Gloo Validation.
Rate Limit Configs are added to the webhook.

`TranslatorSyncerExtensionParams` has been changed so that the `RateLimitServiceSettings` is now a pointer to the `ServiceSettings`.

Bulk validation of resources may be slower or less performant, with the addition and value gained from validating Rate Limit Configs. Please use the list GVK to bulk validate resources in a single API call.

The Validator interface has changed, we have deleted the `ModificationIsSupported`, and `DeletionIsSupported` methods, as they are no longer needed.  All supported resources are now mentioned in the `gateway-validation-webhook-configuration.yaml`.

The `IgnoreProxyValidationFailure` flag was intended to ignore network issues between the Gateway server and the Gloo server. Since the Gateway server and the Gloo server have merged, this is no longer an issue that can occur, and the field is now deprecated.

We have also made suggestions about adding in metrics for validation, such as time spent validating, or time spent locked. These are referenced in the Issue [tagged on this PR](https://github.com/solo-io/gloo/issues/6777#issuecomment-1263872978).

# Context

Adding Extensions to Gloo Validation. This will allow for validation of Rate Limit Configs.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

sam-heilbron commented on projects/gateway/pkg/validation/validator.go:
```diff
@@ -427,6 +427,9 @@ func (v *validator) ValidateModifiedGvk(ctx context.Context, gvk schema.GroupVer
 func (v *validator) validateModifiedResource(ctx context.Context, gvk schema.GroupVersionKind, resource resources.Resource, dryRun, acquireLock bool) (*Reports, error) {
 	var reports *Reports
+	// TODO-JAKE
+	// so it looks like we might need to extend the helm chart in enterprise to add the rate limit resources
```
I'm a little confused by this comment: it refers to helm charts and rate limit resources. Perhaps it will be clearer in the future when you embark on it, but what is the change that will be necessary and why?

jackstine replied:
this comment can be ignored now, because I was able to just add the resource GVK to the helm charts in Gloo.

sam-heilbron replied:
Which GVK did you adjust in the chart? Are you referring to the validation webhook config to accept rate limits?

jackstine replied:
yeah 
https://github.com/solo-io/gloo/blob/7f492fdbb3021bada8ed71552e11c0e489a8b61d/install/helm/gloo/templates/5-gateway-validation-webhook-configuration.yaml#L39

sam-heilbron commented on projects/gloo/api/external/solo/ratelimit/rate_limit_config.go:
```diff
@@ -79,6 +79,7 @@ func (r *RateLimitConfig) GetNamespacedStatuses() *core.NamespacedStatuses {
 }
 
 func (r *RateLimitConfig) SetNamespacedStatuses(status *core.NamespacedStatuses) {
+	contextutils.LoggerFrom(context.Background()).DPanic(status.Statuses)
```
We can sync on this, but this line should never be executed. have you noticed that it is?

jackstine replied:
Yeah this line was/is being called. I have not checked it recently, but it was showing in the logs earlier.

sam-heilbron replied:
"this line was/is being called'- is this true before your changes or just in this PR?

sam-heilbron commented on projects/gloo/pkg/syncer/extauth/extauth_translator_syncer.go:
```diff
@@ -41,6 +42,19 @@ func (s *translatorSyncerExtension) ID() string {
 }
 
+// TODO-JAKE later
+func (s *translatorSyncerExtension) Translate(
```
See the earlier proposal to handle this: Instead of implementing a new method, just re-using Sync but passing parameters that do nothing to the snapshot cache

sam-heilbron commented on projects/gloo/pkg/syncer/ratelimit/ratelimit_translator_syncer.go:
```diff
@@ -39,6 +41,23 @@ func (s *translatorSyncerExtension) ID() string {
 	return ServerRole
 }
 
+func (s *translatorSyncerExtension) Translate(
```
Not required for this PR:

https://github.com/solo-io/gloo/blob/master/projects/gloo/pkg/syncer/ratelimit/ratelimit_translator_syncer.go#L28

I think we could remove this open source syncer extension entirely. Its purpose is to warn users about using enterprise features that they don't have access to. We introduced the [enterprise warning plugin](https://github.com/solo-io/gloo/blob/master/projects/gloo/pkg/plugins/enterprise_warning/plugin.go) to solve this problem. So you could add a rate limit check to that plugin and delete all of this code if you wanted.

jackstine replied:
So it is still being called,  Looks like the [UnmarshalResource](https://github.com/solo-io/solo-kit/blob/master/pkg/utils/protoutils/marshal.go#L188) in solo-kit needs to conform to both CustomInputResources and InputResources.  this can be taken care of with an update.

sam-heilbron reviewed (commented):
Looking good! Left some initial thoughts. Happy to discuss them in person too if you'd like.

Other thought: The webhook API has some fields that we may want to expose to users (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#validatingwebhook-v1-admissionregistration-k8s-io) to make it easier to opt into (or out of) certain fields. We don't need to do it as part of this PR, but I think its important to outline the impact o this new step that occurs during validation, to potentially give users a way to opt out if at first it's not performant enough.

jackstine commented on projects/gloo/pkg/syncer/syncer_extensions.go:
```diff
@@ -30,6 +30,10 @@ type TranslatorSyncerExtension interface {
+
+	// Translate will process the ApiSnapshot and updates the reports with Errors/Warnings that it encouters
+	// unlike Sync() this will not update the SnapshotCache.
+	Translate(ctx context.Context, snap *v1snap.ApiSnapshot, proxies v1.ProxyList, reports reporter.ResourceReports) error
```
No I had not consider using a no-op for the snapshot setter.  A concern though was that the list of proxies came from the snapshot itself.  Since we validate one proxy at a time, I wanted a way to validate a separate proxy individually. This could be done by switching the list of proxies in the snapshot.

jackstine commented on projects/gloo/pkg/validation/gloo_validator.go:
```diff
@@ -5,43 +5,56 @@ import (
 var GvkToSupportedDeleteGlooResources = map[schema.GroupVersionKind]bool{
-	gloov1.UpstreamGVK: true,
-	gloov1.SecretGVK:   true,
+	gloov1.UpstreamGVK:           true,
```
yeah, I was thinking about this as well, we could see about adding the validation for the extensions to the Gateway validator. That way we could check all the proxies at once, after we have checked all the gateway and gloo resources. This would keep the gloo resources separated.

jackstine commented on projects/gloo/pkg/validation/server.go:
```diff
@@ -40,15 +49,23 @@ type validator struct {
 		notifyResync:  make(map[*validation.NotifyOnResyncRequest]chan struct{}, 1),
-		ctx:           ctx,
-		xdsSanitizer:  xdsSanitizer,
+		ctx:           config.Ctx,
```
Not sure how the Ctx is nil?
For your second point, we use ctx in a lot of structs in the setup_syncer. Maybe we could make a refactor.

sam-heilbron replied:
Oh my mistake, I was  confusing the validatorConfig and glooValidatorConfig, one of which has a ctx property and the other does not

jackstine commented on install/helm/gloo/templates/5-gateway-validation-webhook-configuration.yaml:
```diff
@@ -36,6 +36,10 @@ webhooks:
     apiGroups: ["gloo.solo.io"]
     apiVersions: ["v1"]
     resources: ["secrets"]
+  - operations: [ "CREATE", "UPDATE", "DELETE" ]
```
They should apply all their rate limit configs first, before applying vs and route tables that will reference the RLC.  Where is a good place to highlight this information?

kdorosh commented on install/helm/gloo/templates/5-gateway-validation-webhook-configuration.yaml:
```diff
@@ -27,7 +27,11 @@ webhooks:
+  - operations: [ "CREATE", "UPDATE"]
+    apiGroups: ["gateway.solo.io"]
+    apiVersions: ["v1"]
+    resources: ["gateways"]
```
why remove delete for gateways?

jackstine replied:
This was the initial support.  The gateway would not be supported for delete operations, only create/update.

https://github.com/solo-io/gloo/pull/7385#issuecomment-1307323152

jenshu replied:
why did we remove validation of Gateway deletion? wouldn't that be a regression since we used to validate that, but now we don't?

jackstine replied:
deletion of gateways were never supported. 
https://github.com/solo-io/gloo/blob/v1.12.x/projects/gateway/pkg/services/k8sadmission/validating_admission_webhook.go#L401

jenshu replied:
ah gotcha, forgot about that!

kdorosh commented on test/kube2e/gateway/gateway_test.go:
```diff
@@ -2086,14 +2101,65 @@ spec:
+				})
+
+				It("rejects bad resources", func() {
+					testCases := []testCase{{
```
nit; perhaps we can add this as a test case to the existing validation tests? (basically where this was copied from)

jackstine replied:
I separated it because we have to set `EnableExtensionValidation = true`

jackstine commented:
I added in commits [2b5ae741012f32cdb4bb5e0c2f1b298b18c07433](https://github.com/solo-io/gloo/pull/7385/commits/2b5ae741012f32cdb4bb5e0c2f1b298b18c07433) to get rid of the support methods in the validation webhook.  This also updated the helm webhook configuration so that the webhook would only support the options that were initially intended. We can get rid of this commit, if it seems to cause a problem.

Changes to the webhook configuration can be added later in another PR to support all resources.

jackstine commented on projects/gloo/pkg/validation/gloo_validator.go:
```diff
@@ -79,12 +92,22 @@ func (gv glooValidator) Validate(ctx context.Context, proxy *gloov1.Proxy, snaps
+				resourceReports.AddError(proxy, err)
+			}
+		}
+
```
I moved the code to its own validator. This way we validate all the proxies at once for the extensions.

sam-heilbron commented on install/helm/gloo/templates/5-gateway-validation-webhook-configuration.yaml:
```diff
@@ -27,7 +27,11 @@ webhooks:
   - operations: [ "CREATE", "UPDATE", "DELETE" ]
     apiGroups: ["gateway.solo.io"]
     apiVersions: ["v1"]
-    resources: ["*"]
```
In the [gateway API group](https://github.com/solo-io/gloo/tree/master/projects/gateway/api/v1), we have: gateway, rt, vs, and matchable_http_gateway. I'm not sure if we support validating that 4th option in the code, but I did want to call out that this change will mean those matchable http gateways will not be subject to the validation webhook at all. Is that intentional?

jackstine replied:
Should be all good, Here is the [GVK](https://github.com/solo-io/gloo/blob/validate-rate-limit-config/projects/gateway/pkg/api/v1/matchable_http_gateway.sk.go#L175) of the MatchableHttpGateway, which was initially ignored.  See [this comment](https://github.com/solo-io/gloo/pull/7385#discussion_r1014172135) for more details.

sam-heilbron commented on install/helm/gloo/templates/5-gateway-validation-webhook-configuration.yaml:
```diff
@@ -27,7 +27,11 @@ webhooks:
     apiVersions: ["v1"]
-    resources: ["*"]
+    resources: ["virtualservices", "routetables"]
+  - operations: [ "CREATE", "UPDATE"]
```
Given the uncertainty around lack of delete support for gateways, do you think it's worth it to explicitly comment in the code outlining why its not supported?

jackstine replied:
that is fair for now.

sam-heilbron commented on projects/gloo/api/external/solo/ratelimit/rate_limit_config.go:
```diff
@@ -79,6 +79,8 @@ func (r *RateLimitConfig) GetNamespacedStatuses() *core.NamespacedStatuses {
 }
 
 func (r *RateLimitConfig) SetNamespacedStatuses(status *core.NamespacedStatuses) {
+	// TODO-JAKE find out what is calling this, how this is being called.
```
Should we remove this? Have we identified why this code was being executed?

jackstine replied:
just made an update to the comment, we will have to make a change to solo-kit I believe.

sam-heilbron replied:
Let's chat in our sync this morning. The rateLimitConfig is different than most other gloo resources, so I wouldn't expect solo-kit to be required to support this

sam-heilbron commented on projects/gloo/pkg/syncer/envoy_translator_syncer.go:
```diff
@@ -68,6 +68,7 @@ func measureResource(ctx context.Context, resource string, len int) {
 	}
 }
 
+// syncEnvoy this will translate, sanatize, and set the snapshot for each of the proxies, all while merging all the reports into allReports.
```
nit:
```suggestion
// syncEnvoy will translate, sanatize, and set the snapshot for each of the proxies, all while merging all the reports into allReports.
```

sam-heilbron commented on projects/gloo/pkg/syncer/validation/validator.go:
```diff
@@ -0,0 +1,42 @@
+// Validate will sync the extensions with the snapshot. This uses a no-op snapshot setter, so that no changes to the
+// snapshot occur.
+func (v *validator) Validate(ctx context.Context, snapshot *gloosnapshot.ApiSnapshot) reporter.ResourceReports {
+	noOp := &syncer.NoOpSnapshotSetter{}
```
nit: Since this is a no-op, we could instantiate it once on the struct, instead of creating it each time we call validate

sam-heilbron commented on test/kube2e/gateway/gateway_test.go:
```diff
@@ -1252,60 +1292,71 @@ var _ = Describe("Kube2e: gateway", func() {
-	Context("tests with RateLimitConfigs", func() {
-
-		var rateLimitConfig *v1alpha1skv1.RateLimitConfig
+	Context("validation server is disabled", func() {
```
How is the validation server disabled for these tests?

sam-heilbron commented on projects/gloo/pkg/syncer/syncer_extensions.go:
```diff
@@ -34,7 +40,7 @@ type TranslatorSyncerExtension interface {
 type TranslatorSyncerExtensionParams struct {
 	Hasher                   func(resources []envoycache.Resource) (uint64, error)
-	RateLimitServiceSettings ratelimit.ServiceSettings
+	RateLimitServiceSettings *ratelimit.ServiceSettings
```
This is a change to an exported field of an exported struct, so I think it would be considered breaking. Since this code is used internally (ie with the enterprise repo), it's not an issue per se, but I think we should call this out loudly in the PR. Ideally we merge the enterprise code right after this is released, but we should do our best to notify other devs of this change since they may run into it

jackstine replied:
We should make that a practice before we merge PRs.

sam-heilbron commented on projects/gloo/pkg/syncer/setup/setup_syncer.go:
```diff
@@ -772,32 +788,27 @@ func RunGlooWithExtensions(opts bootstrap.Opts, extensions Extensions) error {
-		ignoreProxyValidationFailure,
-		allowWarnings,
-	))
+	syncerValidator := syncerValidation.NewValidator(syncerExtensions, opts.Settings)
```
It may be helpful to outline in a comment what the syncerValidator is, since the setup code is responsible for kicking off so many components. 

Right now we instantiate the validator with the same syncer extensions that are provided at runtime. For enterprise, doesn't this include extauth configs as well? So even though we may be validating a rate limit config, now we'll translate the extauth configs as well. Or am I missing something?

jackstine replied:
yes that is correct.

sam-heilbron commented on projects/gloo/pkg/syncer/setup/setup_syncer.go:
```diff
@@ -772,32 +788,27 @@ func RunGlooWithExtensions(opts bootstrap.Opts, extensions Extensions) error {
-		ignoreProxyValidationFailure,
-		allowWarnings,
-	))
+	syncerValidator := syncerValidation.NewValidator(syncerExtensions, opts.Settings)
```
nit: I think syncerValidator could be made a bit more concrete. We have "syncers" in Gloo Edge, which are responsible for syncing the state of resources. The glooValidator for exampe, validates gloo resources, I think a more appropriate name might be syncerExtensionValidator, or extensionValidator since it's responsible for validating the syncer extensions

sam-heilbron commented on projects/gateway/pkg/validation/validator.go:
```diff
@@ -585,10 +504,22 @@ func (v *validator) getReportsFromGlooValidation(reports []*gloovalidation.GlooV
+	return errs
+}
+
+func (v *validator) getErrorsFromResourceReports(reports reporter.ResourceReports) error {
```
Could we use the `validate` and `validateStrict` methods that are exposed on the ResourceReports https://github.com/solo-io/solo-kit/blob/master/pkg/api/v2/reporter/reporter.go#L196?

sam-heilbron commented on projects/gloo/pkg/syncer/validation/validator.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+type Validator interface {
+	// Validate will validate the proxies of a snapshot, and return the resource reports corresponding to the
```
nit: It's not that we're validating the proxies of the snapshot, we're more validating certain resources (the extended resources) of the snapshot.

sam-heilbron commented on projects/gloo/pkg/syncer/validation/validator.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+// NewValidator will create a new validator for the purpose of validating extensions.
+func NewValidator(extensions []syncer.TranslatorSyncerExtension, settings *v1.Settings) validator {
```
nit: return the reference?

elcasteel commented:
On the performance side, it's important to note that we only validate one resource at a time so validating more types of resources has the potential for a noticeable impact on the performance of applying a lot of config at once. I'm especially thinking of the particular use case where a customer has ratelimit and authconfig for each VS.

elcasteel commented on projects/gloo/pkg/syncer/setup/setup_syncer.go:
```diff
@@ -1106,7 +1117,6 @@ func constructOpts(ctx context.Context, clientset *kubernetes.Interface, kubeCac
 			ValidatingWebhookPort:        gwdefaults.ValidationWebhookBindPort,
 			ValidatingWebhookCertPath:    validationCfg.GetValidationWebhookTlsCert(),
 			ValidatingWebhookKeyPath:     validationCfg.GetValidationWebhookTlsKey(),
-			IgnoreProxyValidationFailure: validationCfg.GetIgnoreGlooValidationFailure(),
```
This was exposed as a setting before right? If so, we should call out in the changelog that it's no longer supported, why and what setting to use instead

jackstine replied:
I can add it to the changelog as well, the docs are covered in these changes.  There is no replacement for this field.

jackstine commented:
Changes that I have made to silence the errors from Ext-Auth.
[Delete the Error for a missing Ext-Auth Config](https://github.com/solo-io/gloo/pull/7385/commits/fe1599609430c0ef02c2d3a588bfb9621ad31237)
Changes made to support changes in tests [one](https://github.com/solo-io/gloo/pull/7385/commits/fb9b69c725e1395332580a82573951c9876294e3) [two](https://github.com/solo-io/gloo/pull/7385/commits/3faf329c904a5891170133a3940232a7009470a4) [three](https://github.com/solo-io/gloo/pull/7385/commits/24cf81d239cc85bf254012c0d2c11bb5cf74b5c3) [four](https://github.com/solo-io/gloo/pull/7385/commits/53840e865f870d181076a8520948a782e2875194)

jackstine commented:
Here is the performance test of applying 800 resources.

Settings:
```
    validation:
      allowWarnings: true
      alwaysAccept: false
      disableTransformationValidation: false
      proxyValidationServerAddr: gloo:9988
      serverEnabled: true
      validationServerGrpcMaxSizeBytes: 104857600
      warnRouteShortCircuiting: false
```

The only errors from the validation were those related to the missing Secret. This is because the applier does not create the secret.

```
objects start:  2022-11-10T15:32:54-06:00
7 / 800 [->_____________________________________________________________________________________________________________________________________________________________] 0.88% ? p/s ?First error encountered at index: 0 admission webhook "gloo.gloo-system.svc" denied the request: resource incompatible with current Gloo snapshot: [Validating *v1.VirtualService failed: 1 error occurred:
	* Validating *v1.VirtualService failed: validating *v1.VirtualService name:"vs-0"  namespace:"gloo-system": failed to validate Proxy [namespace: gloo-system, name: gateway-proxy] with gloo validation: Listener Error: SSLConfigError. Reason: SSL secret not found: list did not find secret gloo-system.client-tls-secret

]
800 / 800 [--------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 3 p/s 5m0s
objects done:  2022-11-10T15:37:54-06:00
```

jackstine commented:
not sure what is causing this issue, but it looks like maybe the upstream group `(Name: test, Namespace: gloo-system)` was never deleted in the previous test?

```
{"level":"error","ts":"2022-11-11T14:21:31.763Z","logger":"gloo.v1.event_loop.setup.gateway-validation-webhook","caller":"k8sadmission/validating_admission_webhook.go:354","msg":"Validation failed: 1 error occurred:\n\t* Validating *v1.VirtualService failed: 1 error occurred:\n\t* Validating *v1.VirtualService failed: validating *v1.VirtualService name:\"missing-upstream\" namespace:\"gloo-system\": failed gloo validation resource reports: 2 errors occurred:\n\t* invalid resource gloo-system.gateway-proxy\n\t* upstream group not found, (Name: test, Namespace: gloo-system)
```

jackstine commented:
/kick

sam-heilbron commented on projects/gloo/pkg/translator/network_filters.go:
```diff
@@ -73,22 +73,6 @@ func (n *httpNetworkFilterTranslator) ComputeNetworkFilters(params plugins.Param
 	// was removed. If we want to support other network filters, we would process
 	// those plugins here.
 
-	// Check that we don't refer to nonexistent auth config
```
How did you confirm/test that the behavior this protected against is still functional?

jackstine replied:
this should only occur in the extensions. This is a link to the test that should hit in Gloo.
https://github.com/solo-io/gloo/blob/10fc47fb95863982bb9a74bb9318d83bcb1785dd/test/kube2e/gateway/gateway_test.go#L2081

 I will add a test in the enterprise to show that this is occurring.

jackstine replied:
brought this back in

jackstine commented on projects/gateway/pkg/validation/validator.go:
```diff
@@ -338,11 +284,7 @@ func (v *validator) validateSnapshot(opts *validationOptions) (*Reports, error)
 		glooReports, err := v.glooValidator(ctx, proxy, opts.Resource, opts.Delete)
 		if err != nil {
 			err = errors.Wrapf(err, failedGlooValidation)
-			if v.ignoreProxyValidationFailure {
-				contextutils.LoggerFrom(ctx).Error(err)
-			} else {
-				errs = multierr.Append(errs, err)
-			}
```
The error that is returned now is if the snapshot has not been received by the gloo validator.
Let me add more context around the deprecated field.
https://github.com/solo-io/gloo/blob/validate-rate-limit-config/projects/gloo/pkg/validation/server.go#L210

The `ignoreProxyValidationFailure` flag was intended to ignore network issues between the Gateway server and the Gloo server.  Since the Gateway server and the Gloo server have merged, this is no longer an issue that can occur, and the field is now deprecated.

kdorosh replied:
i don't think we can remove this, see https://github.com/solo-io/gloo/pull/7385/files#r1024550463

jackstine commented on projects/gateway/pkg/validation/validator.go:
```diff
@@ -585,31 +493,30 @@ func (v *validator) getReportsFromGlooValidation(reports []*gloovalidation.GlooV
-func getErrors(err error) []string {
-	if err == nil {
-		return []string{}
+// UnmarshalResource is the same as the solo-kit pkg/utils/protoutils.Unmarshal() except it does not set the status of the resource
```
let me check that out, and we can make a PR to change solo-kit

nfuden replied:
Can we link the PR?

jackstine replied:
here is the issue 
https://github.com/solo-io/solo-kit/issues/528

jackstine commented on projects/gloo/pkg/translator/translator_test.go:
```diff
@@ -455,7 +455,7 @@ var _ = Describe("Translator", func() {
 
 	Context("Auth configs", func() {
-		It("will error if auth config is missing", func() {
+		It("will not error if auth config is missing", func() {
```
this should only be called out in enterprise.  I can add to the changelog, if you like. But Ext-Auth should not be used in Gloo.  If there is an existing AuthConfig that is referenced, then it will report an error stating that auth config is enterprise only.  If the config is missing, then nothing occurs in Gloo.  Else in Enterprise it will report a warning on the missing Auth Config.

jackstine replied:
Since Auth-Configs are enterprise only, we should not this in the enterprise changelog.

elcasteel replied:
I'm still confused about the reasoning behind this behavior change. Is there a corresponding update in solo-projects to error on missing auth configs? Is there something about the change to validation that required moving that check?

jackstine replied:
Leaving this behavior as it was before

jackstine commented on test/kube2e/gateway/gateway_test.go:
```diff
@@ -1252,60 +1292,71 @@ var _ = Describe("Kube2e: gateway", func() {
-	Context("tests with RateLimitConfigs", func() {
-
-		var rateLimitConfig *v1alpha1skv1.RateLimitConfig
+	Context("validation will not always accept", func() {
```
Yeah the wording is incorrect.

I think there is already enough validation in enterprise.

jackstine commented on changelog/v1.13.0-beta27/extension-validation.yaml:
```diff
@@ -0,0 +1,8 @@
+changelog:
+  - type: NEW_FEATURE
+    description: >
+      Validation of extensions. This can be used to validate Rate Limit Configs and Ext-Auth Configs.
```
I like that idea.

jackstine replied:
I will just say Validation of `TranslatorSyncerExtensions` is now enabled. since that is what we are doing.

And move the other resources to Enterprise Changelog.

jackstine commented on projects/gateway/pkg/validation/validator.go:
```diff
@@ -533,48 +474,15 @@ func (v *validator) validateResource(opts *validationOptions) (*Reports, error)
+// getErrorsFromGlooValidation returns an error comprising of the gloo reports. The errors will include warnings if
+// allowWarnings is not set.
+func (v *validator) getErrorsFromGlooValidation(reports []*gloovalidation.GlooValidationReport) error {
+	var errs error
```
I am sorry I do not follow.
1. what is the "Report type"?

nfuden replied:
https://github.com/solo-io/gloo/pull/7385/files/5023798220783e2cdbea6d78bf004677e9e444ec#diff-abf1ea0c5351ae08b421977226f5aeddce902eb5caba0aa9beef54837a68bc6eR482 
The report that error came from I assume

elcasteel commented on projects/gateway/pkg/services/k8sadmission/validating_admission_webhook.go:
```diff
@@ -397,9 +395,16 @@ func (wh *gatewayValidationWebhook) validateAdmissionRequest(
 	}
+
+	if _, hit := gloosnapshot.ApiGvkToHashableResource[gvk]; !hit {
+		contextutils.LoggerFrom(ctx).Debugf("unsupported validation for resource namespace [%s] name [%s] group [%s] kind [%s]", ref.GetNamespace(), ref.GetName(), gvk.Group, gvk.Kind)
```
If users are looking at logs from an aggregator rather than directly on the pod, they don't always have previous lines to provide context. As a general principle I like to treat this kind of debug message as something that needs to stand alone. (which is to say leave this code as is, assuming everything here is helpful for understanding why a this resource wasn't validated)

elcasteel commented on projects/gateway/pkg/validation/validator_test.go:
```diff
@@ -10,14 +10,16 @@ import (
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 	"github.com/solo-io/gloo/pkg/utils"
-	gatewayv1 "github.com/solo-io/gloo/projects/gateway/pkg/api/v1"
```
thanks for cleaning this up!

kdorosh commented on changelog/v1.13.0-beta27/extension-validation.yaml:
```diff
@@ -0,0 +1,10 @@
+changelog:
+  - type: NEW_FEATURE
+    description: >
+      Validation of `TranslatorSyncerExtensions` is now enabled.
```
since the changelog is user facing and read by our customers, i prefer this reads as it will affect them. more technical details you have written here (e.g. interface changes) should go in the PR description.

i think a user really just cares that we added support for ratelimit and extauth validation

kdorosh commented on projects/gateway/pkg/validation/validator.go:
```diff
@@ -338,11 +284,7 @@ func (v *validator) validateSnapshot(opts *validationOptions) (*Reports, error)
 		glooReports, err := v.glooValidator(ctx, proxy, opts.Resource, opts.Delete)
 		if err != nil {
 			err = errors.Wrapf(err, failedGlooValidation)
-			if v.ignoreProxyValidationFailure {
```
I don't think we can remove this. we still support deploying the gateway and gloo pods (even if the default changed) and until we remove that functionality we should not remove this config

elcasteel replied:
We don't support deploying the gateway pod at all. We support user supplied proxies but this validation webhook is never making a network call.

kdorosh replied:
oh interesting, ok nevermind then i like this change. thanks for sharing @elcasteel

jackstine replied:
How can you deploy the gateway and gloo pods?

elcasteel replied:
there is never a gateway pod (post 1.12)

kdorosh commented on test/kube2e/gateway/gateway_test.go:
```diff
@@ -77,6 +76,46 @@ var _ = Describe("Kube2e: gateway", func() {
+			placeholderUs.Metadata.Name = fmt.Sprintf("invalid-placeholder-us-%d", attempt)
+
+			_, err := resourceClientset.UpstreamClient().Write(placeholderUs, clients.WriteOpts{Ctx: ctx})
+			if err != nil {
+				// We have successfully rejected an invalid upstream
+				// This means that the webhook is fully warmed, and contains a Snapshot with a Proxy
+				return true
```
this can also happen if the error is a network error. we should ensure the error message is a validation rejection by checking the error message

kdorosh commented on test/kube2e/gateway/gateway_test.go:
```diff
@@ -77,6 +76,46 @@ var _ = Describe("Kube2e: gateway", func() {
+		}
+		attempt := 0
+		Eventually(func(g Gomega) bool {
+			placeholderUs.Metadata.Name = fmt.Sprintf("invalid-placeholder-us-%d", attempt)
```
nit; you probably want a newline here

kdorosh reviewed (commented):
lgtm overall!

jackstine commented:
[xds-relay test is failing](https://github.com/solo-io/gloo/actions/runs/3491260446/jobs/5843688618)

nfuden commented on changelog/v1.13.0-beta27/extension-validation.yaml:
```diff
@@ -0,0 +1,7 @@
+changelog:
+  - type: NEW_FEATURE
+    description: >
+      Validation webhook will now validate Auth Configs and Rate Limit Configs.
```
are we always validating them? Is this togglable?

jackstine replied:
Validated only if referenced on a VS or Route, is that some detail that we should include?

jackstine commented:
Ext-Auth configs in the following commits
https://github.com/solo-io/gloo/pull/7385/commits/417aab48e46b06a521a1319dfefe879f4c86994e
https://github.com/solo-io/gloo/pull/7385/commits/f5244f3b957ff86e3f11b8a5f7ea76eac84810bc
